cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.h

```diff
@@ -0,0 +1,497 @@
+        std::vector<double> diffusion_activation_volume;
+        std::vector<double> diffusion_creep_prefactor;
+        std::vector<double> diffusion_creep_grain_size_exponent;
+
+        /**
+         * Reference viscosity profile coordinates, and the corresponding viscosity.
+         */
+        std::vector<double> reference_viscosity_coordinates;
+        std::unique_ptr<Rheology::AsciiDepthProfile<dim>> reference_viscosity_profile;
+
+        /**
+         * A reference profile for density scaling.
+         */
+        Utilities::AsciiDataProfile<dim> rho_vs_depth_profile;
+
+        /**
+         * The column indices of the density scaling column in the ascii profile file.
+         */
+        unsigned int density_scaling_index;
+
+
+        /**
+         * A reference profile for the thermal expansivity.
+         */
+        Utilities::AsciiDataProfile<dim> thermal_expansivity_profile;
+        unsigned int thermal_expansivity_column_index;
+
+        /**
+         * A reference profile for temperatura scaling. Values are from Steinberger and
+         * Calderwood, 2006.
+         */
+        Utilities::AsciiDataProfile<dim> dT_vs_depth_profile;
+        /**
+         * The column indices of the density scaling column in the ascii profile file.
+         */
+        unsigned int temperature_scaling_index;
+
+        /**
+         * An object of ascii data boundary to input crustal depths.
+         */
+        Utilities::AsciiDataBoundary<dim> crustal_boundary_depth;
+
+        /**
+         * A depth-profile of the laterally averaged viscosity in each layer
+         * in the current model. Can be used to compare (and potentially scale)
+         * the computed viscosity to the reference profile.
+         */
+        std::vector<double> average_viscosity_profile;
+
+        /**
+         * Variable returned  to determine if the evaluate () funciton is called and
+         * viscosities are computed. Initially, it is set to false and then updated
+         * to true in the update () funciton.
+         */
+        bool initialized;
+
+        /**
+         * Because of the nonlinear nature of this material model many
+         * parameters need to be kept within bounds to ensure stability of the
+         * solution. These bounds can be adjusted as input parameters.
+         */
+        double max_temperature_dependence_of_eta;
+        double min_eta;
+        double max_eta;
+        double min_specific_heat;
+        double max_specific_heat;
+        double min_thermal_expansivity;
+        double max_thermal_expansivity;
+        unsigned int max_latent_heat_substeps;
+        double min_grain_size;
+
+        double diffusion_viscosity (const double      temperature,
+                                    const double      pressure,
+                                    const std::vector<double>    &compositional_fields,
+                                    const SymmetricTensor<2,dim> &,
```

Comment:
Can you give this argument a name? Or do you not actually use it in the implementation? In the latter case, just remove the argument altogether.